File: prolog/triangle/triangle.pl
%degenerate
triangle(S1, S2, S3, _) :- S1 + S2 < S3, !, fail.
triangle(S1, S2, S3, _) :- S2 + S3 < S1, !, fail.
triangle(S1, S2, S3, _) :- S3 + S1 < S2, !, fail.
%zero
triangle(S1, S2, S3, _) :- S1 = 0, S2 = 0, S3 = 0, !, fail.

triangle(S1, S2, S3, "equilateral") :- S1 = S2, S2 = S3, S2 = S1, !.
triangle(S1, S2, S3, "scalene") :- S1 \= S2, S2 \= S3, S2 \= S1, !.
triangle(S, S, _, "isosceles") :- !.
triangle(S, _, S, "isosceles") :- !.
triangle(_, S, S, "isosceles") :- !.




%mine shame
triangle(Side1, Side2, Side3, Type) :- is_triangle(Side1, Side2, Side3),
    ((Type = "equilateral", is_equilateral(Side1, Side2, Side3));
    (Type = "isosceles", is_isosceles(Side1, Side2, Side3));
    (Type = "scalene", is_scalene(Side1, Side2, Side3))).

is_equilateral(Side1, Side2, Side3) :- Side1 =:= Side2, Side1 =:= Side3.

is_isosceles(Side1, Side2, Side3) :-
    (Side1 =:= Side2, Side1 =\= Side3);
    (Side1 =:= Side3, Side1 =\= Side2);
    (Side2 =:= Side3, Side2 =\= Side1).

is_scalene(Side1, Side2, Side3) :-
    (Side1 =\= Side2, Side1 =\= Side3, Side2 =\= Side3).

is_triangle(Side1, Side2, Side3) :-
    Side1 + Side2 >= Side3,
    Side2 + Side3 >= Side1,
    Side1 + Side3 >= Side2,
    Side1 =\= 0.

File: solutions/python/triangle/2/triangle.py
def equilateral(sides):
    a, b, c = sides
    return False if not is_triangle(sides) \
        else a is b and a is c and b is c


def isosceles(sides):
    a, b, c = sides

    if not is_triangle(sides):
        return False

    if equilateral(sides):
        return True

    return (a == b and a != c) \
        or (a != b and a == c) \
        or (b == c and a != c)


def scalene(sides):
    a, b, c = sides

    return is_triangle(sides) and a is not b and b is not c and a is not c


def is_triangle(sides):
    a, b, c = sides
    if a == 0 and b == 0 and c == 0:
        return False;

    return a + b >= c \
        and b + c >= a \
        and a + c >= b
